source/magpie/operation.d

```diff
@@ -0,0 +1,337 @@
+                indx[j] = to!string(combinator.indx.row.codes[j][i]);
+            }
+            else
+            {
+                indx[j] = combinator.indx.row.index[j][combinator.indx.row.codes[j][i]];
+            }
+        }
+
+        return indx;
+    }
+
+    static if(type == "left")
+    {
+        combinator.indx.row = df1.indx.row;
+        static foreach(i; 0 .. df1.RowType.length)
+            combinator.data[i] = df1.data[i];
+        static foreach(i; 0 .. df2.RowType.length)
+            combinator.data[df1.RowType.length + i].length = df1.rows;
+
+        foreach(i; 0 .. combinator.indx.row.codes[0].length)
+        {
+            string[] indx = unpack(i);
+            const int pos = df2.indx.getPosition!0(indx);
+
+            if(pos > -1)
+            {
+                static foreach(j; 0 .. df2.RowType.length)
+                {
+                    combinator.data[df1.RowType.length + j][i] = df2.data[j][pos];
+                }
+            }
+        }
+
+        combinator.rows = df1.rows;
+        return combinator;
+    }
+    else static if(type == "right")
+    {
+        combinator.indx.row = df2.indx.row;
+        static foreach(i; 0 .. df2.RowType.length)
+            combinator.data[df1.RowType.length + i] = df2.data[i];
+        static foreach(i; 0 .. df1.RowType.length)
+            combinator.data[i].length = df2.rows;
+
+        foreach(i; 0 .. combinator.indx.row.codes[0].length)
+        {
+            string[] indx = unpack(i);
+            const int pos = df1.indx.getPosition!0(indx);
+
+            if(pos > -1)
+            {
+                static foreach(j; 0 .. df1.RowType.length)
+                    combinator.data[j][i] = df1.data[j][pos];
+            }
+        }
+
+        combinator.rows = df2.rows;
+        return combinator;
+    }
+    else static if(type == "outer")
+    {
+        import magpie.helper: indexUnion;
+        combinator.indx.row.index = transposed(indexUnion!0(df1.indx, df2.indx));
+        combinator.indx.row.titles = df1.indx.row.titles;
+
+        combinator.indx.generateCodes();
+        combinator.indx.optimize();
+        combinator.rows = combinator.indx.row.codes[0].length;
+
+        static foreach(i; 0 .. combinator.RowType.length)
+            combinator.data[i].length = combinator.rows;
+
+        foreach(i; 0 .. combinator.rows)
+        {
+            string[] indx = unpack(i);
+
+            int pos = df1.indx.getPosition!0(indx);
+            if(pos > -1)
+            {
+                static foreach(j; 0 .. df1.RowType.length)
+                    combinator.data[j][i] = df1.data[j][pos];
+            }
+
+            pos = df2.indx.getPosition!0(indx);
+            if(pos > -1)
+            {
+                static foreach(j; 0 .. df2.RowType.length)
+                    combinator.data[df1.RowType.length + j][i] = df2.data[j][pos];
+            }
+        }
+
+        return combinator;
+    }
+    else static if(type == "inner" || type == "")
```

Comment:
just set the default value of `type` to `"inner"`

Reply:
Addressed in 6517b4a